workflows/methylation/methylation-preprocess.wdl

```diff
@@ -0,0 +1,209 @@
+            annotation: "Annotation object",
+            beta: "Beta values",
+            cn_values: "Copy number values",
+            m_values: "M values",
+            pheno: "Phenotype data",
+            pheno_data: "Phenotype data",
+            probe_names: "Probe names",
+            sample_names: "Sample names",
+        }
+    }
+
+    parameter_meta {
+        idats: "Array of raw IDAT files from the Illumina methlyation array"
+    }
+
+    input {
+        Pair[File, File] idats
+    }
+
+    call process_raw_idats { input:
+        idats,
+    }
+
+    #@ except: LineWidth
+    output {
+        File beta_swan_norm_unfiltered = process_raw_idats.beta_swan_norm_unfiltered
+        File beta_swan_norm_unfiltered_genomic = process_raw_idats.beta_swan_norm_unfiltered_genomic
+        File mset = process_raw_idats.mset
+        File rgset = process_raw_idats.rgset
+        File rset = process_raw_idats.rset
+        File annotation = process_raw_idats.annotation
+        File beta = process_raw_idats.beta
+        File cn_values = process_raw_idats.cn_values
+        File m_values = process_raw_idats.m_values
+        File pheno = process_raw_idats.pheno
+        File pheno_data = process_raw_idats.pheno_data
+        File probe_names = process_raw_idats.probe_names
+        File sample_names = process_raw_idats.sample_names
+    }
+}
+
+task process_raw_idats {
+    meta {
+        description: "Process raw IDAT files from the Illumina methylation array for a single sample"
+        outputs: {
+            beta_swan_norm_unfiltered: "Normalized beta values for all probes",
+            beta_swan_norm_unfiltered_genomic: "Normalized beta values for all probes that map to the genome",
+            mset: "MethylSet object",
+            rgset: "RGSet object",
+            rset: "RatioSet object",
+            annotation: "Annotation object",
+            beta: "Beta values",
+            cn_values: "Copy number values",
+            m_values: "M values",
+            pheno: "Phenotype data",
+            pheno_data: "Phenotype data",
+            probe_names: "Probe names",
+            sample_names: "Sample names",
+        }
+    }
+
+    parameter_meta {
+        idats: "Array of raw IDAT files from the Illumina methlyation array"
+        disk_size_gb: "Disk size in GB"
+    }
+
+    input {
+        Pair[File, File] idats
+        Int disk_size_gb = 10
+    }
+
+    String idat_base = basename(idats.left)
+    String out_base = sub(idat_base, "(_Grn|_Red).idat", "")
+
+    #@ except: LineWidth
+    command <<<
+        echo "Processing IDAT files"
+        ln -s ~{idats.left} ~{idats.right} .
+
+        R --no-save <<SCRIPT
+            options(error = function() traceback(3))
+
+            library(minfi)
+            library(dplyr)
+            library(data.table)
+            library(IlluminaHumanMethylationEPICmanifest)
+            library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
+
+            set.seed(1)
```

Comment:
Is there ever a reason for setting another seed?

Reply:
These arrays have two types of probes (helpfully Infinium type 1 and Infinium type 2). The probes also cover a varying number of CpG sites. The normalization method works by taking `N` probes of each type that each have 1, 2, or 3 CpGs. So it selects 6N probes in total. This selection is "random" if the seed is not fixed, of course. Since we're processing samples individually in this step instead of as a cohort, I want to seed to be consistent so the same set of probes is chosen for each sample.

Reply:
Makes sense. Can you add that to the documentation somewhere? Maybe it should go in `meta.help`? Or just an embedded "normal" comment, not sure. Your call.